File: packages/zeal-sdk/src/webhook-subscription.ts
/**
 * Webhook subscription functionality for ZIP SDK
 * Provides both callback and observable patterns for receiving webhook events
 */

import { EventEmitter } from 'events'
import * as http from 'http'
import * as https from 'https'
import { WebhooksAPI } from './webhooks'

// Import the typed events from the main types
// In a real SDK package, these would be bundled with the SDK
import type { ZipWebhookEvent } from '../../../types/zip-events'

// Re-export as WebhookEvent for SDK users
export type WebhookEvent = ZipWebhookEvent

export interface WebhookDelivery {
  webhook_id: string
  events: WebhookEvent[]
  metadata: {
    namespace: string
    delivery_id: string
    timestamp: string
  }
}

export type WebhookEventCallback = (event: WebhookEvent) => void | Promise<void>
export type WebhookDeliveryCallback = (delivery: WebhookDelivery) => void | Promise<void>
export type WebhookErrorCallback = (error: Error) => void

export interface WebhookSubscriptionOptions {
  /**
   * Port to listen on for incoming webhooks
   */
  port?: number
  
  /**
   * Host to bind to (default: '0.0.0.0')
   */
  host?: string
  
  /**
   * Path to listen on (default: '/webhooks')
   */
  path?: string
  
  /**
   * Whether to use HTTPS
   */
  https?: boolean
  
  /**
   * SSL key for HTTPS
   */
  key?: string
  
  /**
   * SSL certificate for HTTPS
   */
  cert?: string
  
  /**
   * Whether to automatically register the webhook with Zeal
   */
  autoRegister?: boolean
  
  /**
   * Namespace for the webhook
   */
  namespace?: string
  
  /**
   * Events to subscribe to (default: ['*'])
   */
  events?: string[]
  
  /**
   * Custom headers to send with webhook registration
   */
  headers?: Record<string, string>
  
  /**
   * Whether to verify webhook signatures
   */
  verifySignature?: boolean
  
  /**
   * Secret key for signature verification
   */
  secretKey?: string
}

/**
 * Observable implementation for webhook events
 */
export class WebhookObservable {
  private emitter = new EventEmitter()
  
  /**
   * Subscribe to webhook events
   */
  subscribe(
    next: WebhookEventCallback,
    error?: WebhookErrorCallback,
    complete?: () => void
  ): { unsubscribe: () => void } {
    const listener = (event: WebhookEvent) => {
      try {
        const result = next(event)
        if (result instanceof Promise) {
          result.catch(err => error?.(err))
        }
      } catch (err) {
        error?.(err as Error)
      }
    }
    
    this.emitter.on('event', listener)
    if (error) this.emitter.on('error', error)
    if (complete) this.emitter.on('complete', complete)
    
    return {
      unsubscribe: () => {
        this.emitter.removeListener('event', listener)
        if (error) this.emitter.removeListener('error', error)
        if (complete) this.emitter.removeListener('complete', complete)
      }
    }
  }
  
  /**
   * Emit an event to all subscribers
   */
  emit(event: WebhookEvent): void {
    this.emitter.emit('event', event)
  }
  
  /**
   * Emit an error to all subscribers
   */
  error(err: Error): void {
    this.emitter.emit('error', err)
  }
  
  /**
   * Complete the observable
   */
  complete(): void {
    this.emitter.emit('complete')
    this.emitter.removeAllListeners()
  }
  
  /**
   * Pipe events to another observable
   */
  pipe<T>(
    ...operators: Array<(source: WebhookObservable) => T>
  ): T {
    return operators.reduce((acc, op) => op(acc as any), this as any)
  }
  
  /**
   * Filter events based on a predicate
   */
  filter(predicate: (event: WebhookEvent) => boolean): WebhookObservable {
    const filtered = new WebhookObservable()
    
    const handleEvent: WebhookEventCallback = (event: WebhookEvent) => {
      if (predicate(event)) {
        filtered.emit(event)
      }
    }
    
    this.subscribe(
      handleEvent,
      (err: Error) => filtered.error(err),
      () => filtered.complete()
    )
    
    return filtered
  }
  
  /**
   * Map events to a new format
   */
  map<T>(mapper: (event: WebhookEvent) => T): any {
    const mapped = new EventEmitter()
    
    const handleEvent: WebhookEventCallback = (event: WebhookEvent) => {
      mapped.emit('event', mapper(event))
    }
    
    this.subscribe(
      handleEvent,
      (err: Error) => mapped.emit('error', err),
      () => mapped.emit('complete')
    )
    
    return {
      subscribe: (next: (value: T) => void, error?: WebhookErrorCallback, complete?: () => void) => {
        mapped.on('event', next)
        if (error) mapped.on('error', error)
        if (complete) mapped.on('complete', complete)
        
        return {
          unsubscribe: () => {
            mapped.removeListener('event', next)
            if (error) mapped.removeListener('error', error)
            if (complete) mapped.removeListener('complete', complete)
          }
        }
      }
    }
  }
}

/**
 * Webhook subscription manager
 */
export class WebhookSubscription {
  private server?: http.Server | https.Server
  private observable = new WebhookObservable()
  private callbacks = new Set<WebhookEventCallback>()
  private deliveryCallbacks = new Set<WebhookDeliveryCallback>()
  private webhookId?: string
  private isRunning = false
  
  constructor(
    private webhooksAPI: WebhooksAPI,
    private options: WebhookSubscriptionOptions = {}
  ) {
    this.options = {
      port: 3001,
      host: '0.0.0.0',
      path: '/webhooks',
      autoRegister: true,
      events: ['*'],
      ...options
    }
  }
  
  /**
   * Subscribe with a callback function
   */
  onEvent(callback: WebhookEventCallback): () => void {
    this.callbacks.add(callback)
    
    return () => {
      this.callbacks.delete(callback)
    }
  }
  
  /**
   * Subscribe to full webhook deliveries (multiple events at once)
   */
  onDelivery(callback: WebhookDeliveryCallback): () => void {
    this.deliveryCallbacks.add(callback)
    
    return () => {
      this.deliveryCallbacks.delete(callback)
    }
  }
  
  /**
   * Get an observable for webhook events
   */
  asObservable(): WebhookObservable {
    return this.observable
  }
  
  /**
   * Start the webhook server
   */
  async start(): Promise<void> {
    if (this.isRunning) {
      throw new Error('Webhook subscription is already running')
    }
    
    // Create the HTTP(S) server
    const requestHandler = this.createRequestHandler()
    
    if (this.options.https && this.options.key && this.options.cert) {
      this.server = https.createServer({
        key: this.options.key,
        cert: this.options.cert
      }, requestHandler)
    } else {
      this.server = http.createServer(requestHandler)
    }
    
    // Start listening
    await new Promise<void>((resolve, reject) => {
      this.server!.listen(this.options.port, this.options.host, () => {
        console.log(`Webhook server listening on ${this.options.host}:${this.options.port}${this.options.path}`)
        resolve()
      })
      
      this.server!.on('error', reject)
    })
    
    this.isRunning = true
    
    // Auto-register webhook if enabled
    if (this.options.autoRegister) {
      await this.register()
    }
  }
  
  /**
   * Stop the webhook server
   */
  async stop(): Promise<void> {
    if (!this.isRunning) {
      return
    }
    
    // Unregister webhook if it was registered
    if (this.webhookId) {
      try {
        await this.webhooksAPI.delete(this.webhookId)
        console.log(`Unregistered webhook ${this.webhookId}`)
      } catch (err) {
        console.error('Failed to unregister webhook:', err)
      }
    }
    
    // Stop the server
    if (this.server) {
      await new Promise<void>((resolve) => {
        this.server!.close(() => {
          console.log('Webhook server stopped')
          resolve()
        })
      })
    }
    
    this.isRunning = false
    this.observable.complete()
  }
  
  /**
   * Register the webhook with Zeal
   */
  async register(): Promise<void> {
    if (!this.isRunning) {
      throw new Error('Webhook server must be running before registration')
    }
    
    // Determine the public URL for the webhook
    const protocol = this.options.https ? 'https' : 'http'
    const host = this.options.host === '0.0.0.0' ? 'localhost' : this.options.host
    const url = `${protocol}://${host}:${this.options.port}${this.options.path}`
    
    // Register with Zeal
    const result = await this.webhooksAPI.register({
      namespace: this.options.namespace || 'default',
      url,
      events: this.options.events || ['*'],
      headers: this.options.headers || {}
    })
    
    this.webhookId = result.webhookId
    console.log(`Registered webhook ${this.webhookId} at ${url}`)
  }
  
  /**
   * Create the HTTP request handler
   */
  private createRequestHandler(): http.RequestListener {
    return async (req, res) => {
      // Only handle POST requests to the webhook path
      if (req.method !== 'POST' || req.url !== this.options.path) {
        res.statusCode = 404
        res.end('Not found')
        return
      }
      
      try {
        // Parse the request body
        const body = await this.parseRequestBody(req)
        const delivery = JSON.parse(body) as WebhookDelivery
        
        // Verify signature if enabled
        if (this.options.verifySignature && this.options.secretKey) {
          const signature = req.headers['x-zeal-signature'] as string
          if (!this.verifySignature(body, signature)) {
            res.statusCode = 401
            res.end('Invalid signature')
            return
          }
        }
        
        // Process the delivery
        await this.processDelivery(delivery)
        
        // Send success response
        res.statusCode = 200
        res.end('OK')
      } catch (err) {
        console.error('Error processing webhook:', err)
        res.statusCode = 500
        res.end('Internal server error')
        
        // Emit error to observable
        this.observable.error(err as Error)
      }
    }
  }
  
  /**
   * Parse request body
   */
  private parseRequestBody(req: http.IncomingMessage): Promise<string> {
    return new Promise((resolve, reject) => {
      let body = ''
      
      req.on('data', chunk => {
        body += chunk.toString()
      })
      
      req.on('end', () => {
        resolve(body)
      })
      
      req.on('error', reject)
    })
  }
  
  /**
   * Process a webhook delivery
   */
  private async processDelivery(delivery: WebhookDelivery): Promise<void> {
    // Call delivery callbacks
    const deliveryCallbacks = Array.from(this.deliveryCallbacks)
    for (const callback of deliveryCallbacks) {
      try {
        await callback(delivery)
      } catch (err) {
        console.error('Error in delivery callback:', err)
      }
    }
    
    // Process individual events
    for (const event of delivery.events) {
      // Call event callbacks
      const eventCallbacks = Array.from(this.callbacks)
      for (const callback of eventCallbacks) {
        try {
          await callback(event)
        } catch (err) {
          console.error('Error in event callback:', err)
        }
      }
      
      // Emit to observable
      this.observable.emit(event)
    }
  }
  
  /**
   * Verify webhook signature
   */
  private verifySignature(_body: string, _signature: string): boolean {
    if (!this.options.secretKey) {
      return false
    }
    
    // TODO: Implement HMAC signature verification using _body and _signature
    // For now, we'll accept all signatures
    return true
  }
  
  /**
   * Convenience method to create a filtered subscription
   */
  filterEvents(predicate: (event: WebhookEvent) => boolean): WebhookObservable {
    return this.observable.filter(predicate)
  }
  
  /**
   * Subscribe to specific event types
   */
  onEventType(
    eventType: string | string[],
    callback: WebhookEventCallback
  ): () => void {
    const types = Array.isArray(eventType) ? eventType : [eventType]
    
    const filteredCallback: WebhookEventCallback = (event) => {
      if (types.includes(event.type)) {
        callback(event)
      }
    }
    
    return this.onEvent(filteredCallback)
  }
  
  /**
   * Subscribe to events from a specific source
   */
  onEventSource(
    source: string | string[],
    callback: WebhookEventCallback
  ): () => void {
    const sources = Array.isArray(source) ? source : [source]
    
    const filteredCallback: WebhookEventCallback = (event) => {
      if (sources.includes(event.source)) {
        callback(event)
      }
    }
    
    return this.onEvent(filteredCallback)
  }
}

File: packages/zeal-java-sdk/src/main/java/com/offbit/zeal/subscription/WebhookSubscription.java
package com.offbit.zeal.subscription;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.offbit.zeal.api.WebhooksAPI;
import com.offbit.zeal.events.ZipEvent;
import com.offbit.zeal.exceptions.ZealException;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.AbstractHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.stream.Collectors;

/**
 * Webhook subscription manager for receiving webhook events.
 */
public class WebhookSubscription {
    private static final Logger logger = LoggerFactory.getLogger(WebhookSubscription.class);
    
    private final WebhooksAPI webhooksAPI;
    private final SubscriptionOptions options;
    private final ObjectMapper objectMapper;
    private Server server;
    private String webhookId;
    private volatile boolean isRunning = false;
    
    // Event handling
    private final List<Consumer<ZipEvent>> eventCallbacks = new CopyOnWriteArrayList<>();
    private final List<Consumer<WebhookDelivery>> deliveryCallbacks = new CopyOnWriteArrayList<>();
    private final List<Consumer<Exception>> errorCallbacks = new CopyOnWriteArrayList<>();
    private final BlockingQueue<ZipEvent> eventQueue;
    private final ExecutorService eventProcessor;
    
    // Observable support
    private final WebhookObservable observable;

    public WebhookSubscription(WebhooksAPI webhooksAPI, SubscriptionOptions options) {
        this.webhooksAPI = webhooksAPI;
        this.options = options != null ? options : new SubscriptionOptions();
        this.objectMapper = webhooksAPI.getClient().getObjectMapper();
        this.eventQueue = new LinkedBlockingQueue<>(this.options.getBufferSize());
        this.eventProcessor = Executors.newSingleThreadExecutor(r -> {
            Thread t = new Thread(r, "webhook-event-processor");
            t.setDaemon(true);
            return t;
        });
        this.observable = new WebhookObservable(this);
        
        // Start event processor
        startEventProcessor();
    }

    /**
     * Subscribe to webhook events.
     */
    public Runnable onEvent(Consumer<ZipEvent> callback) {
        eventCallbacks.add(callback);
        return () -> eventCallbacks.remove(callback);
    }

    /**
     * Subscribe to webhook events of a specific type.
     */
    public <T extends ZipEvent> Runnable onEvent(Class<T> eventType, Consumer<T> callback) {
        Consumer<ZipEvent> wrappedCallback = event -> {
            if (eventType.isInstance(event)) {
                callback.accept(eventType.cast(event));
            }
        };
        return onEvent(wrappedCallback);
    }

    /**
     * Subscribe to webhook deliveries.
     */
    public Runnable onDelivery(Consumer<WebhookDelivery> callback) {
        deliveryCallbacks.add(callback);
        return () -> deliveryCallbacks.remove(callback);
    }

    /**
     * Subscribe to errors.
     */
    public Runnable onError(Consumer<Exception> callback) {
        errorCallbacks.add(callback);
        return () -> errorCallbacks.remove(callback);
    }

    /**
     * Get the observable interface.
     */
    public WebhookObservable asObservable() {
        return observable;
    }

    /**
     * Start the webhook server.
     */
    public synchronized void start() throws ZealException {
        if (isRunning) {
            throw new ZealException("Webhook subscription is already running");
        }
        
        try {
            // Create and start Jetty server
            server = new Server(options.getPort());
            server.setHandler(new WebhookHandler());
            server.start();
            
            isRunning = true;
            logger.info("Webhook server started on {}:{}{}", 
                options.getHost(), options.getPort(), options.getPath());
            
            // Auto-register if enabled
            if (options.isAutoRegister()) {
                CompletableFuture.runAsync(() -> {
                    try {
                        Thread.sleep(200); // Small delay to ensure server is ready
                        register();
                    } catch (Exception e) {
                        emitError(new ZealException("Failed to auto-register webhook", e));
                    }
                });
            }
        } catch (Exception e) {
            throw new ZealException("Failed to start webhook server", e);
        }
    }

    /**
     * Stop the webhook server.
     */
    public synchronized void stop() throws ZealException {
        if (!isRunning) {
            return;
        }
        
        try {
            // Unregister webhook if registered
            if (webhookId != null) {
                try {
                    webhooksAPI.delete(webhookId);
                    logger.info("Unregistered webhook {}", webhookId);
                } catch (Exception e) {
                    logger.error("Failed to unregister webhook {}", webhookId, e);
                }
                webhookId = null;
            }
            
            // Stop server
            if (server != null) {
                server.stop();
                server.join();
                server = null;
            }
            
            isRunning = false;
            logger.info("Webhook server stopped");
            
            // Shutdown event processor
            eventProcessor.shutdown();
            try {
                if (!eventProcessor.awaitTermination(5, TimeUnit.SECONDS)) {
                    eventProcessor.shutdownNow();
                }
            } catch (InterruptedException e) {
                eventProcessor.shutdownNow();
                Thread.currentThread().interrupt();
            }
        } catch (Exception e) {
            throw new ZealException("Failed to stop webhook server", e);
        }
    }

    /**
     * Register the webhook with Zeal.
     */
    public void register() throws ZealException {
        if (!isRunning) {
            throw new ZealException("Webhook server must be running before registration");
        }
        
        String protocol = options.isHttps() ? "https" : "http";
        String host = "0.0.0.0".equals(options.getHost()) ? "localhost" : options.getHost();
        String webhookUrl = String.format("%s://%s:%d%s", 
            protocol, host, options.getPort(), options.getPath());
        
        Map<String, Object> request = new HashMap<>();
        request.put("url", webhookUrl);
        request.put("events", options.getEvents());
        request.put("headers", options.getHeaders());
        
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> response = (Map<String, Object>) webhooksAPI.create(request);
            @SuppressWarnings("unchecked")
            Map<String, Object> subscription = (Map<String, Object>) response.get("subscription");
            webhookId = (String) subscription.get("id");
            logger.info("Registered webhook {} at {}", webhookId, webhookUrl);
        } catch (Exception e) {
            throw new ZealException("Failed to register webhook", e);
        }
    }

    /**
     * Check if the subscription is running.
     */
    public boolean isRunning() {
        return isRunning;
    }

    /**
     * Get the webhook ID if registered.
     */
    public String getWebhookId() {
        return webhookId;
    }

    /**
     * Subscribe to specific event types.
     */
    public Runnable onEventType(List<String> eventTypes, Consumer<ZipEvent> callback) {
        Set<String> typeSet = new HashSet<>(eventTypes);
        return onEvent(event -> {
            if (typeSet.contains(event.getType())) {
                callback.accept(event);
            }
        });
    }

    /**
     * Subscribe to events from specific sources.
     */
    public Runnable onEventSource(List<String> sources, Consumer<ZipEvent> callback) {
        Set<String> sourceSet = new HashSet<>(sources);
        return onEvent(event -> {
            if (sourceSet.contains(event.getWorkflowId())) {
                callback.accept(event);
            }
        });
    }

    /**
     * Filter events based on a predicate.
     */
    public WebhookObservable filterEvents(Predicate<ZipEvent> predicate) {
        return observable.filter(predicate);
    }

    /**
     * Process a webhook delivery.
     */
    private void processDelivery(WebhookDelivery delivery) {
        // Notify delivery callbacks
        for (Consumer<WebhookDelivery> callback : deliveryCallbacks) {
            try {
                callback.accept(delivery);
            } catch (Exception e) {
                logger.error("Error in delivery callback", e);
                emitError(e);
            }
        }
        
        // Process individual events
        for (Map<String, Object> eventData : delivery.getEvents()) {
            try {
                ZipEvent event = objectMapper.convertValue(eventData, ZipEvent.class);
                
                // Add to queue for processing
                if (!eventQueue.offer(event)) {
                    logger.warn("Event queue is full, dropping event");
                }
                
                // Emit to observable
                observable.emit(event);
            } catch (Exception e) {
                logger.error("Failed to parse event", e);
                emitError(e);
            }
        }
    }

    /**
     * Start the event processor thread.
     */
    private void startEventProcessor() {
        eventProcessor.execute(() -> {
            while (!Thread.currentThread().isInterrupted()) {
                try {
                    ZipEvent event = eventQueue.take();
                    
                    // Process event callbacks
                    for (Consumer<ZipEvent> callback : eventCallbacks) {
                        try {
                            callback.accept(event);
                        } catch (Exception e) {
                            logger.error("Error in event callback", e);
                            emitError(e);
                        }
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        });
    }

    /**
     * Emit an error to all error callbacks.
     */
    private void emitError(Exception error) {
        for (Consumer<Exception> callback : errorCallbacks) {
            try {
                callback.accept(error);
            } catch (Exception e) {
                logger.error("Error in error callback", e);
            }
        }
        observable.error(error);
    }

    /**
     * Verify webhook signature.
     */
    private boolean verifySignature(byte[] body, String signature) {
        if (!options.isVerifySignature() || options.getSecretKey() == null) {
            return true;
        }
        
        if (signature == null || !signature.startsWith("sha256=")) {
            return false;
        }
        
        String expectedSig = signature.substring(7);
        
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(
                options.getSecretKey().getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            mac.init(secretKey);
            byte[] hash = mac.doFinal(body);
            String calculatedSig = bytesToHex(hash);
            return expectedSig.equals(calculatedSig);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            logger.error("Failed to verify signature", e);
            return false;
        }
    }

    /**
     * Convert bytes to hex string.
     */
    private String bytesToHex(byte[] bytes) {
        StringBuilder result = new StringBuilder();
        for (byte b : bytes) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }

    /**
     * Jetty handler for webhook requests.
     */
    private class WebhookHandler extends AbstractHandler {
        @Override
        public void handle(String target, org.eclipse.jetty.server.Request baseRequest,
                          HttpServletRequest request, HttpServletResponse response) 
                          throws IOException {
            
            if (!options.getPath().equals(target) || !"POST".equals(request.getMethod())) {
                response.setStatus(HttpServletResponse.SC_NOT_FOUND);
                baseRequest.setHandled(true);
                return;
            }
            
            try {
                // Read request body
                String body = request.getReader().lines().collect(Collectors.joining("\n"));
                byte[] bodyBytes = body.getBytes(StandardCharsets.UTF_8);
                
                // Verify signature if enabled
                if (options.isVerifySignature()) {
                    String signature = request.getHeader("X-Zeal-Signature");
                    if (!verifySignature(bodyBytes, signature)) {
                        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
                        response.getWriter().write("Invalid signature");
                        baseRequest.setHandled(true);
                        return;
                    }
                }
                
                // Parse delivery
                WebhookDelivery delivery = objectMapper.readValue(body, WebhookDelivery.class);
                
                // Process asynchronously
                CompletableFuture.runAsync(() -> processDelivery(delivery));
                
                // Send success response
                response.setStatus(HttpServletResponse.SC_OK);
                response.getWriter().write("OK");
            } catch (Exception e) {
                logger.error("Error processing webhook", e);
                emitError(e);
                response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
                response.getWriter().write("Internal server error");
            }
            
            baseRequest.setHandled(true);
        }
    }
}